src/passes/GUFA.cpp

```diff
@@ -0,0 +1,311 @@
+        replaceCurrent(getDroppedChildrenAndAppend(
+          curr, wasm, options, builder.makeUnreachable()));
+      } else {
+        // We can't remove this, but we can at least put an unreachable
+        // right after it.
+        replaceCurrent(builder.makeSequence(builder.makeDrop(curr),
+                                            builder.makeUnreachable()));
+      }
+      optimized = true;
+    };
+
+    if (contents.getType() == Type::unreachable) {
+      // This cannot contain any possible value at all. It must be unreachable
+      // code.
+      replaceWithUnreachable();
+      return;
+    }
+
+    // This is reachable. Check if we can emit something optimized for it.
+    // TODO: can we handle more general things here too?
+    if (!contents.canMakeExpression()) {
+      return;
+    }
+
+    if (contents.isNull() && curr->type.isNullable()) {
+      // Null values are all identical, so just fix up the type here if we need
+      // to (the null's type might not fit in this expression, if it passed
+      // through casts).
+      if (!Type::isSubType(contents.getType(), curr->type)) {
+        contents = PossibleContents::literal(
+          Literal::makeNull(curr->type.getHeapType()));
+      }
+
+      // Note that if curr's type is *not* nullable, then the code will trap at
+      // runtime (the null must arrive through a cast that will trap). We handle
+      // that below, so we don't need to think about it here.
+
+      // TODO: would emitting a more specific null be useful when valid?
+    }
+
+    auto* c = contents.makeExpression(wasm);
+
+    // We can only place the constant value here if it has the right type. For
+    // example, a block may return (ref any), that is, not allow a null, but in
+    // practice only a null may flow there if it goes through casts that will
+    // trap at runtime.
+    // TODO: GUFA should eventually do this, but it will require it properly
+    //       filtering content not just on ref.cast as it does now, but also
+    //       ref.as etc. Once it does those we could assert on the type being
+    //       valid here.
+    if (Type::isSubType(c->type, curr->type)) {
+      if (canRemove(curr)) {
+        replaceCurrent(getDroppedChildrenAndAppend(curr, wasm, options, c));
+      } else {
+        // We can't remove this, but we can at least drop it and put the
+        // optimized value right after it.
+        replaceCurrent(builder.makeSequence(builder.makeDrop(curr), c));
+      }
+      optimized = true;
+    } else {
+      // The type is not compatible: we cannot place |c| in this location, even
+      // though we have proven it is the only value possible here.
+      if (Properties::isConstantExpression(c)) {
+        // The type is not compatible and this is a simple constant expression
+        // like a ref.func. That means this code must be unreachable. (See below
+        // for the case of a non-constant.)
+        replaceWithUnreachable();
```

Comment:
The casts would be along the way:
```wat
(block (result (ref data))
  (ref.as_data
    (ref.func $foo)))
```
The contents `c` here are a `ref.func $foo`. But the type at the block is `data` which is impossible. The cast along the way allowed the contents to get here in the GUFA flow, and also they allowed the wasm to validate, but at runtime the cast will trap.

In general, if we see the contents must have a type that is not valid for the declared type in the wasm, then either we have a bug, or the wasm spec has a bug, or the code is unreachable, since the wasm type system should prevent this situation.

Reply:
Ah right, sorry, I keep confusing the original expression with the expression created by the contents oracle...